---
title: "j2-1 Stoichiometry of lysine hydroxylations"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteIndexEntry{Bioconductor style for PDF documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   html_document:
     highlight: haddock
     toc: yes
     toc_depth: 2
     keep_md: yes
---


# Environment setup


```{r load packages, message = FALSE, warning = FALSE}

temp <- sapply(list.files("../functions", full.names = TRUE), source)

processors <- 8

ind.fa.dir <- file.path("../../data/processed_data/individual_fastas")
results.dir <- normalizePath(file.path("../../results"))


j2.res.dir <- file.path(results.dir, "j2-PTM-stoichiometry")

create.dirs(c(
    j2.res.dir
))

```


# Data import


```{r fasta data}

cells <- c(
    "HeLa_WT", "HeLa_J6KO",
    "HEK293", "MCF7",
    "HeLa_WT_J6PD", "HeLa_J6KO_J6PD", "HeLa_WT_J6PD_2"
)

all.pp.dt <- lapply(
    cells,
    function(x){
        dt <- fread(
            file.path(
                "../../data/rawdata/DB_search_psm/",
                paste0(x, "__protein-peptides.csv")
            )
        )
        setnames(
            dt,
            old = c(
                "Protein Accession",
                grep("Area", colnames(dt), value = TRUE),
                "#Feature"
                ),
            new = c("Accession", "Area", "n_feature")
        )
        dt[, `:=`(
            cell = x,
            Peptide = gsub("(^[[:alpha:]]\\.|\\.[[:alpha:]]$)", "", Peptide)
        )]
        dt[, .(cell, Accession, Start, End, Peptide, Area, PTM, n_feature)]
    }
) %>% rbindlist

all.pp.dt[, cell := factor(cell, levels = cells)]

all.iupred2.dt <- fread(file.path(j.a.res.dir, "IUPRED2_per_protein_and_position.csv"))
k.ratio.dt <- fread(file.path(j.a.res.dir, "k_ratio_per_protein_and_position.csv"))

```


# Data filtration

## Preliminary checks


```{r preliminary checks}

all.dt <- copy(all.pp.dt)
print("All K related modifications")
all.letters <- all.dt[, unlist(str_split(Peptide, "(?=[[:alpha:]])"))]
all.letters <- all.letters[all.letters != ""]
table(all.letters[grepl("^K", all.letters)])
all.dt[, aa_list := str_split(Peptide, "(?=[[:alpha:]])")]

print("All PTMs")
PTMs <- all.dt[, unlist(str_split(PTM, "; "))]
PTMs <- PTMs[PTMs != ""]
table(sort(PTMs))

```

## Filtration

```{r data filtration}

filterDt <- function(all.dt){
    
    print(all.dt[, table(cell)])

    all.dt <- all.dt[
        Area > 0 &
        Accession %in% names(all.protein.bs)
    ]

    print(all.dt[, table(cell)])

    ## Remove Acetylation from modification
    all.dt[, corrected_peptide := gsub("\\(\\+42\\.01\\)", "", Peptide)]

    ## Filter out peptides with C-term propionylation and peptides which do not have C-term K, KOH or R
    all.dt <- all.dt[
        str_count(corrected_peptide, "K\\(\\+56\\.03\\)$") +
        str_count(corrected_peptide, "K\\(\\+72\\.02\\)$") == 0 &
        (
            str_count(corrected_peptide, "R$") == 1 |
            str_count(corrected_peptide, "(K$|K\\(\\+15\\.99\\)$)") == 1 
        )
    ]

    print(all.dt[, table(cell)])

    ## Also K and K(+15.99) for at most one occurence in the peptide (with ignoring C terminus K and KP)
    all.dt[
      , Peptide_for_filtration := gsub(
            "(K\\(\\+15\\.99\\)$|K$|R$)", "", corrected_peptide # C-term K or R is OK
        ) %>%        
            {gsub("K\\(\\+56\\.03\\)", "X", .)} %>%
            {gsub("K\\(\\+72\\.02\\)", "X", .)} %>%
            {gsub("\\s*\\([^\\)]+\\)", "", .)} %>%
            {gsub("KP", "XZ", .)} %>%
            {gsub("RP", "XZ", .)}
    ]

    all.dt <- all.dt[str_count(Peptide_for_filtration, "(K|R)") < 2]
    all.dt[, Peptide_for_filtration := NULL]

    print(all.dt[, table(cell)])

    all.dt[, PTMs := str_split(PTM, "; ")]

    all.dt <- all.dt[
      , corrected_peptide := case_when(
            !any(PTMs[[1]] == "Oxidation (K)") ~
                gsub("K\\(\\+15.99\\)","K", corrected_peptide),
            TRUE ~ corrected_peptide
        ) %>% {case_when(
                   !any(PTMs[[1]] == "Oxidised Propionylation") ~
                       gsub("K\\(\\+72.02\\)","K", .),
                   TRUE ~ .
               )},
        by = seq_len(nrow(all.dt))
    ]

    all.dt[, `:=`(
        raw_peptide = gsub("\\s*\\([^\\)]+\\)","", Peptide),
        k_info_peptide =
            gsub(
                names(AMINO_ACID_CODE) %>%
                {.[. != "K"]} %>%
                {paste(., collapse = "|")} %>%
                {paste0("(", ., ")")}, 
                "A", corrected_peptide # Non K amino acids will be replaced with A
            ) %>%
            {gsub("K\\(\\+56.03\\)", "K", .)} %>% 
            {gsub("(K\\(\\+15.99\\)|K\\(\\+72.02\\))", "H", .)} %>% # H: hydroxylared K
            {gsub("\\s*\\([^\\)]+\\)","", .)}
    )]

    all.dt[, `:=`(
        K_index = str_locate_all(k_info_peptide, "(K|H)") %>%
            {lapply(., function(x) x[,2])},
        oxK_index = str_locate_all(k_info_peptide, "H") %>%
            {lapply(., function(x) x[,2])},
        all_aa_index = str_locate_all(k_info_peptide, "(K|H|A)") %>%
            {lapply(., function(x) x[,2])}
    )]
}

all.pp.dt <- filterDt(all.pp.dt)


```

# Individual amino acid position


```{r index to aa pos}

indexToPosition <- function(index.name, dt){
    dt <- copy(dt)
    dt[, peptide_id := 1:.N]

    dt[
      , n_index := length(get(paste0(index.name, "_index"))[[1]]),
        by = seq_len(nrow(dt)) 
    ]

    e.dt <- dt[rep(1:nrow(dt), times = n_index)]
    e.dt[, index_pos_id := 1:.N, by = peptide_id]
    e.dt[
      , position_in_peptide :=
            get(paste0(index.name, "_index"))[[1]][index_pos_id],
        by = seq_len(nrow(e.dt))
    ]

    e.dt[, `:=`(
        position = Start + position_in_peptide - 1,
        aa_type = index.name
    )]

    e.dt <- e.dt[, c(
        "cell", "Accession", "aa_type", "position", "Area", "n_feature"
    ), with = FALSE]

    pos.dt <- e.dt[, list(
        total_count = .N,
        total_area = sum(Area),
        total_n_feature = sum(n_feature)
    ), by = list(cell, Accession, aa_type, position)]

    return(pos.dt)
}


all.pos.dt <- lapply(
    c("K", "oxK", "all_aa"),
    indexToPosition,
    dt = all.pp.dt
) %>%
    rbindlist


```



# Analysis of stoichiometry and other features


```{r calculate stoic}

stoic.dt <- dcast(
    all.pos.dt[aa_type %in% c("K", "oxK")],
    formula = cell + Accession + position ~ aa_type,
    value.var = c("total_area", "total_n_feature"),
    fill = 0
)

stoic.dt[, `:=`(
    ox_ratio = total_area_oxK / total_area_K
)]

fwrite(
    stoic.dt,
    file.path(j2.res.dir, "hydroxylation_stoichiometry_summary.csv")
)

n.feature.th <- 0
stoic.dt <- stoic.dt[total_n_feature_K > n.feature.th]

stoic.dt[, uniprot_id := str_split_fixed(Accession, "\\|", n = 2)[, 1]]

```

## IDRs

```{r load disorderd score, fig.width = 5}


stoic.dt <- merge(
    stoic.dt,
    all.iupred2.dt[, .(Accession, uniprot_id, position, residue, IUPRED2)],
    by = c("Accession", "uniprot_id", "position")
)

stoic.dt[, `:=`(
    IUPRED2_int = cut(IUPRED2, breaks = c(0, 0.25, 0.5, 0.75, 1)),
    ox_ratio_int = round(ox_ratio, digits = 1)
)]

ggplot(
    data = stoic.dt[grepl("HeLa", cell)],
    aes(
        x = IUPRED2,
        y = ox_ratio,
    )
) +
    geom_point() +
    facet_grid( ~ cell) +
    theme_classic(12) +
    theme(
        legend.title = element_text(),
        legend.position = "bottom",
        aspect.ratio = 1
    ) +
    ggsci::scale_fill_aaas() +
    scale_y_continuous(labels = scales::percent_format()) +
    ylab("Proportion") +
    ggtitle("All detected proteins")


ggplot(
    data = stoic.dt[grepl("HeLa", cell)],
    aes(
        fill = IUPRED2_int,
        x = ox_ratio,
        group = IUPRED2_int
    )
) +
    geom_histogram(
        position = "identity", 
        aes(y = stat(density) * 0.1), binwidth = 0.1
    ) +
    facet_grid(IUPRED2_int ~ cell) +
    theme_classic(12) +
    theme(
        legend.title = element_text(),
        legend.position = "bottom"
    ) +
    ggsci::scale_fill_aaas() +
    coord_cartesian(ylim = c(0, 0.05)) +
    scale_y_continuous(labels = scales::percent_format()) +
    ylab("Proportion") +
    ggtitle("All detected proteins")

ggplot(
    data = stoic.dt[grepl("HeLa", cell) & grepl("(BRD2|BRD3|BRD4)", Accession)],
    aes(
        fill = IUPRED2_int,
        x = ox_ratio,
        group = IUPRED2_int
    )
) +
    geom_histogram(
        position = "identity", 
        aes(y = stat(density) * 0.1), binwidth = 0.1
    ) +
    facet_grid(IUPRED2_int ~ cell) +
    theme_classic(12) +
    theme(
        legend.title = element_text(),
        legend.position = "bottom"
    ) +
    ggsci::scale_fill_aaas() +
    coord_cartesian(ylim = c(0, 0.05)) +
    scale_y_continuous(labels = scales::percent_format()) +
    ylab("Proportion") +
    ggtitle("BRD2-3s only")


```


## K ratio


```{r k ratio}

stoic.dt <- merge(
    stoic.dt,
    k.ratio.dt[, .(Accession, uniprot_id, position, K_ratio, K_ratio_score)],
    by = c("Accession", "uniprot_id", "position")
)

```



```{r plot K ratio, fig.width = 5}

stoic.dt[, K_ratio_int := cut(K_ratio, breaks = c(0, 0.2, 0.4, 0.6, 0.8, 1))]

ggplot(
    data = stoic.dt[grepl("HeLa", cell)],
    aes(
        x = factor(K_ratio),
        y = ox_ratio,
        fill = cell
    )
) +
    geom_boxplot(outlier.shape = NA) +
    theme_classic(12) +
    theme(aspect.ratio = 1) +
    khroma::scale_fill_bright()

ggplot(
    data = stoic.dt[grepl("HeLa", cell)],
    aes(
        fill = K_ratio_int,
        x = ox_ratio,
        group = K_ratio_int
    )
) +
    geom_histogram(
        position = "identity", 
        aes(y = stat(density) * 0.1), binwidth = 0.1
    ) +
    facet_grid(K_ratio_int ~ cell) +
    theme_classic(12) +
    theme(
        legend.title = element_text(),
        legend.position = "bottom"
    ) +
    ggsci::scale_fill_aaas() +
    ##coord_cartesian(ylim = c(0, 0.1)) +
    scale_y_continuous(labels = scales::percent_format()) +
    ylab("Proportion") +
    ggtitle("All detected proteins")


ggplot(
    data = stoic.dt[grepl("HeLa", cell) & grepl("(BRD2|BRD3|BRD4)", Accession)],
    aes(
        fill = K_ratio_int,
        x = ox_ratio,
        group = K_ratio_int
    )
) +
    geom_histogram(
        position = "identity", 
        aes(y = stat(density) * 0.1), binwidth = 0.1
    ) +
    facet_grid(K_ratio_int ~ cell) +
    theme_classic(12) +
    theme(
        legend.title = element_text(),
        legend.position = "bottom"
    ) +
    ggsci::scale_fill_aaas() +
    ##coord_cartesian(ylim = c(0, 0.1)) +
    scale_y_continuous(labels = scales::percent_format()) +
    ylab("Proportion") +
    ggtitle("BRD2-3s")



```


```{r stoic data export}

fwrite(
    stoic.dt,
    file.path(j2.res.dir, "hydroxylation_stoichiometry_summary_extended.csv")
)

wide.dt <- dcast(
    stoic.dt,
    Accession + position + uniprot_id + IUPRED2 + K_ratio + K_ratio_score ~ cell,
    value.var = c("ox_ratio", "total_n_feature_K", "total_n_feature_oxK")
)

library("BSgenome")

wide.dt[, `:=`(
    protein_sequence = BSgenome::getSeq(
                             x = all.protein.bs,
                             name = Accession
                         ) %>% as.character
)]

wide.dt[, `:=`(
    sequence = substr(protein_sequence, start = pmax(0, position - 10), stop = position + 10),
    protein_sequence = NULL
)]

left.cols <- c("Accession", "uniprot_id", "position", "sequence")

wide.dt <- wide.dt[, c(
    left.cols, colnames(wide.dt)[!(colnames(wide.dt) %in% left.cols)]
), with = FALSE]

fwrite(
    wide.dt,
    file.path(j2.res.dir, "hydroxylation_stoichiometry_summary_extended_wide.csv")
)


```



# Session information


```{r session info}

sessioninfo::session_info()


```
