---
title: "j2-3 Analysis of the stoichiometry of lysine hydroxylation and arginine methylation"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteIndexEntry{Bioconductor style for PDF documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   html_document:
     highlight: haddock
     toc: yes
     toc_depth: 2
     keep_md: yes
---


# Environment setup


```{r load packages, message = FALSE, warning = FALSE}

temp <- sapply(list.files("../functions", full.names = TRUE), source)
temp <- sapply(list.files("./functions", full.names = TRUE), source)
library("Biostrings")

processors <- 8

j2.input.dir <- file.path("../../data/j2-PTM-stoichiometry")
j2.1.input.dir <- file.path(j2.input.dir, "hydroxyK-stoichiometry")

results.dir <- normalizePath(file.path("../../results"))
j0.res.dir <- file.path(results.dir, "j0-data-preprocessing")
j2.res.dir <- file.path(results.dir, "j2-PTM-stoichiometry")
j2.3.res.dir <- file.path(j2.res.dir, "j2-3-PTM-stoichiometry-K-and-R")

create.dirs(c(
    j2.res.dir,
    j2.3.res.dir
))

```


# Data import


```{r fasta data}

data.source <- c(
    "HeLa_WT_JQ1",
    "HeLa_JMJD6KO_JQ1",
    "HeLa_WT_J6pep",
    "HeLa_JMJD6KO_J6pep"
)

all.pp.dt<- lapply(
    1:length(data.source),
    function(x){
        dt <- fread(
            paste0(
                j2.3.input.dir,
                "/",
                data.source[x], "_Trypsin_oxKandMetR__protein-peptides.csv"
            )
        )
        setnames(
            dt,
            old = c(
                "Protein Accession",
                grep("Area", colnames(dt), value = TRUE),
                "#Feature"
                ),
            new = c("Accession", "Area", "n_feature")
        )
        dt[, `:=`(
            data_source = data.source[x],
            Peptide = gsub("(^[[:alpha:]]\\.|\\.[[:alpha:]]$)", "", Peptide)
        )]
        dt[, .(data_source, Accession, Start, End, Unique, Peptide, Area, PTM, n_feature)]
    }
) %>% rbindlist

all.pp.dt[, data_source := factor(data_source, levels = data.source)]

print("The number of peptides per data source")
all.pp.dt[, .N, by = data_source]

all.protein.feature.per.pos.dt <- file.path(
    j0.res.dir, "all_protein_feature_per_position.csv"
) %>% fread

```


# Data filtration

## Preliminary checks

```{r preliminary checks}

temp <- countPTM(all.pp.dt)

```

## Filtration and data processing

The following filtration and data processing will be performed:

1. Filter out non-unique peptide
2. Filter out peptides with Area == 0
3. Ignore acetylation
   - (e.g.) Treat K(+42.01) as K or K(+42.01)(+15.99) as K(+15.99)
4. Filter out peptides (1) with C-term propionylated K or dimethylated R and (2) without C-term K, hydroxyK, R or mono-methylated R
   - di-methylated R is not cleaved by trypsin 
5. Allow only 1 misclevage by trypsin 
   - At most one occurrence of K, K(+15.99), R, R(+14.02) in the peptide
   - Note that the following are allowed in the peptides
	 - C terminal K, R, KP, RP
	 - Propionylated K: K(+56.03), K(+72.02)
	 - Dimethyl R: R(+28.03)
6. The PTMs are assigned in the PTM column
7. (additional filter) K does not have M or W within 2 amino acids to be assigned as hydroxy K

```{r data filtration}

all.indexed.pp.dt <- filterPeptideData(all.pp.dt)

all.pos.dt <- lapply(
    c("K", "oxK", "R", "mR", "dmR", "all_aa"),
    indexToPosition,
    dt = all.indexed.pp.dt
) %>%
    rbindlist

```



# Analysis of stoichiometry and other features

## K stoichiometry

```{r calculate stoic}

k.stoic.dt <- dcast(
    all.pos.dt[aa_type %in% c("K", "oxK")],
    formula = data_source + Accession + position ~ aa_type,
    value.var = c("total_area", "total_n_feature"),
    fill = 0
)

k.stoic.dt[, `:=`(
    oxK_ratio = total_area_oxK / total_area_K
)]

all.protein.bs <- readAAStringSet("../../data/all_protein.fasta")
names(all.protein.bs) <- names(all.protein.bs) %>%
    {gsub("sp\\|", "", .)} %>%
    {str_split_fixed(., " ", n = 2)[, 1]}
uniprot.accession.flag <- names(all.protein.bs) %>%
    {str_split_fixed(., "\\|", n = 2)[, 2]} %>%
    {nchar(.) > 0}
all.protein.bs <- all.protein.bs[uniprot.accession.flag]

a.range <- 5

k.stoic.dt[
    , seq5 := BSgenome::getSeq(
                          x = all.protein.bs,
                          name = Accession
                        ) %>% as.character %>%
              substr(start = position - a.range, stop = position + a.range)
]

## print("Before filtration 7")
## k.stoic.dt[, table(Hydroxy_K = oxK_ratio > 0, data_source)]

## MW filter
k.stoic.dt[, `:=`(
    MW_within_1 = substr(
        seq5,
        start = min(a.range, position - 1),
        stop = min(a.range, position - 1) + 2
    ) %>%
        str_count(pattern = "(M|W)") > 0,
    MW_within_2 = substr(
        seq5,
        start = min(a.range, position - 2),
        stop = min(a.range, position - 2) + 4
    ) %>%
        str_count(pattern = "(M|W)") > 0    
), by = seq_len(nrow(k.stoic.dt))]

## Import curated data
curated.hydroxK.site.dt <- fread(
    file.path(j2.input.dir, "curated_JMJD6_substrate_sites.csv") 
)

curated.hydroxK.site.dt[, `:=`(
    curated_oxK_site = TRUE
)]

curated.hydroxK.site.dt <- curated.hydroxK.site.dt[, .(
    Accession, position, curated_oxK_site, screen
)]

all.protein.feature.per.pos.dt <- merge(
    curated.hydroxK.site.dt,
    all.protein.feature.per.pos.dt,
    all.y = TRUE,
    by = c("Accession", "position")
)

all.protein.feature.per.pos.dt[, `:=`(
    curated_oxK_site = case_when(
        is.na(curated_oxK_site) ~ FALSE,
        TRUE ~ curated_oxK_site
    )
)]

merge(
    all.protein.feature.per.pos.dt,
    k.stoic.dt,
    by = c("Accession", "position")
) %>%
    fwrite(file.path(j2.3.res.dir, "long_K_stoichiometry_data.csv"))

wide.k.stoic.dt <- dcast(
    k.stoic.dt,
    formula = Accession + position + seq5 + MW_within_1 + MW_within_2 ~ data_source,
    value.var = c(
        "oxK_ratio",
        grep("^total_area", colnames(k.stoic.dt), value = TRUE),
        grep("^total_n_feature", colnames(k.stoic.dt), value = TRUE)
    )
)

merge(
    all.protein.feature.per.pos.dt,
    wide.k.stoic.dt,
    by = c("Accession", "position")
) %>%
    fwrite(
        file.path(j2.3.res.dir, "wide_K_stoichiometry_data.csv")
    )

```

## R stoichiometry

```{r calculate R stoic}

stoic.dt <- dcast(
    all.pos.dt[aa_type %in% c("R", "mR", "dmR")],
    formula = data_source + Accession + position ~ aa_type,
    value.var = c("total_area", "total_n_feature"),
    fill = 0
)

stoic.dt[, `:=`(
    mR_ratio = total_area_mR / total_area_R,
    dmR_ratio = total_area_dmR / total_area_R
)]

stoic.dt[
    , seq5 := BSgenome::getSeq(
                          x = all.protein.bs,
                          name = Accession
                        ) %>% as.character %>%
              substr(start = position - 5, stop = position + 5)
]

merge(
    all.protein.feature.per.pos.dt,
    stoic.dt,
    by = c("Accession", "position")
) %>%
    fwrite(file.path(j2.3.res.dir, "long_R_stoichiometry_data.csv"))

wide.stoic.dt <- dcast(
    stoic.dt,
    formula = Accession + position + seq5 ~ data_source,
    value.var = c(
        "mR_ratio", "dmR_ratio",
        grep("^total_area", colnames(stoic.dt), value = TRUE),
        grep("^total_n_feature", colnames(stoic.dt), value = TRUE)
    )
)

merge(
    all.protein.feature.per.pos.dt,
    wide.stoic.dt,
    by = c("Accession", "position")
) %>%
    fwrite(
        file.path(j2.3.res.dir, "wide_R_stoichiometry_data.csv")
    )

```


# Session information


```{r session info}

sessioninfo::session_info()


```
