---
title: "Analysis of lysine hydroxylation stoichiometry and protein features"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteIndexEntry{Bioconductor style for PDF documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   html_document:
     highlight: haddock
     toc: yes
     toc_depth: 2
     keep_md: yes
---


# Package import


```{r load packages, message = FALSE, warning = FALSE}

temp <- sapply(list.files("../functions", full.names = TRUE), source)

library("Biostrings")
library("GenomicRanges")

processors <- 8

set.seed(1)

```


```{r dir create}

ind.fa.dir <- file.path("../../data/processed_data/individual_fastas")
results.dir <- normalizePath(file.path("../../results"))

j.a.res.dir <- file.path(results.dir, "j-a-data-preprocessing")
j.b.res.dir <- file.path(results.dir, "j-b-stoichiometry")

create.dirs(c(
))

```


# Data import


```{r fasta data}

all.protein.bs <- readAAStringSet("../../data/all_protein.fasta")
names(all.protein.bs) <- names(all.protein.bs) %>%
    {gsub("sp\\|", "", .)} %>%
    {str_split_fixed(., " ", n = 2)[, 1]}
uniprot.accession.flag <- names(all.protein.bs) %>%
    {str_split_fixed(., "\\|", n = 2)[, 2]} %>%
    {nchar(.) > 0}
all.protein.bs <- all.protein.bs[uniprot.accession.flag]

cells <- c("HeLa_WT", "HeLa_J6KO", "HEK293", "MCF7", "HeLa_WT_J6PD", "HeLa_J6KO_J6PD", "HeLa_WT_J6PD_2")

library("readxl")


manually.curated.hydroxylation.dt <-
    file.path(
        "../../data/processed_data/20210721_JMJD6_manually_curated_hydroxylation_site.xlsx"
    ) %>%
    read_excel %>%
    data.table

manually.curated.hydroxylation.dt <-
    manually.curated.hydroxylation.dt[, .(Accession, position)]
manually.curated.hydroxylation.dt[, JMJD6_substrate_flag := "JMJD6_substrate"]

stoichiometry.dt <- fread(
    file.path(j.b.res.dir, "hydroxylation_stoichiometry_summary_extended.csv")
)

stoichiometry.dt <- merge(
    manually.curated.hydroxylation.dt,
    stoichiometry.dt,
    all.y = TRUE,
    by = c("Accession", "position")
)

stoichiometry.dt[, JMJD6_substrate_flag := case_when(
               is.na(JMJD6_substrate_flag) ~ "Others",
               TRUE ~ JMJD6_substrate_flag
           )]

all.biophysic.dt <- fread(file.path(j.a.res.dir, "biophysical_propety.csv"))
setnames(all.biophysic.dt, old = "Position", new = "position")

stoichiometry.dt <- merge(
    all.biophysic.dt,
    stoichiometry.dt,
    all.y = TRUE,
    by = c("Accession", "position")
)


library("BSgenome")

stoichiometry.dt[, `:=`(
    protein_sequence = BSgenome::getSeq(
                             x = all.protein.bs,
                             name = Accession
                         ) %>% as.character
)]

seq.search.range <- 10

stoichiometry.dt[, `:=`(
    sequence = substr(
        protein_sequence,
        start = pmax(0, position - seq.search.range),
        stop = position + seq.search.range
    ),
    protein_sequence = NULL
)]

stoichiometry.dt[, full_length_seq_flag := (
    nchar(sequence) == (seq.search.range * 2 + 1)
)]

left.cols <- c(
    "cell", "Accession", "uniprot_id", "position", "sequence", "full_length_seq_flag",
    "JMJD6_substrate_flag",
    "ox_ratio"
)

stoichiometry.dt <- stoichiometry.dt[, c(
    left.cols, colnames(stoichiometry.dt)[!(colnames(stoichiometry.dt) %in% left.cols)]
), with = FALSE]

stoichiometry.dt <- stoichiometry.dt[
    residue == "K" &
    total_n_feature_K > 3
]

```

# Analysis of sequence feature around hydroxylation site

## Amino acid enrichment around the hydroxylation sites


```{r amino acid enrichment, fig.width = 7}

non.duplicated.stoichiometry.dt <- stoichiometry.dt[
    full_length_seq_flag == TRUE &
    cell %in% c("HeLa_WT", "HeLa_WT_J6PD", "HeLa_WT_J6PD_2")
][order(ox_ratio, decreasing = TRUE)][
    !duplicated(paste(Accession, position))
]

non.duplicated.stoichiometry.dt[
  , (paste0("position_", gsub("-", "m", seq(-10, 10)))) :=
        str_split(sequence, pattern = "")[[1]] %>% as.list,
    by = seq_len(nrow(non.duplicated.stoichiometry.dt))
]

key.id.cols <- c("cell", "Accession", "position", "JMJD6_substrate_flag", "ox_ratio")

m.non.duplicated.stoichiometry.dt <- melt(
    non.duplicated.stoichiometry.dt,
    id.vars = c(key.id.cols),
    measure.vars = paste0("position_", gsub("-", "m", seq(-10, 10))),
    variable.name = "position2KOH",
    value.name = "amino_acid"
)

aa.count.per.pos.dt <- m.non.duplicated.stoichiometry.dt[
  , .N, by = list(JMJD6_substrate_flag, position2KOH, amino_acid)
]

aa.count.per.pos.dt[
  , total_AA_per_pos := sum(N),
    by = list(JMJD6_substrate_flag, position2KOH)
]

aa.count.per.pos.dt[
  , odds := (N / total_AA_per_pos) %>% {./(1 - .)}
]

odds.ratio.dt <- dcast(
    aa.count.per.pos.dt,
    position2KOH + amino_acid ~ JMJD6_substrate_flag,
    value.var = c("odds", "N", "total_AA_per_pos")
)

setnafill(
    odds.ratio.dt, fill = 0, cols = c("odds_JMJD6_substrate", "N_JMJD6_substrate")
)

odds.ratio.dt[, total_AA_per_pos_JMJD6_substrate := case_when(
                    is.na(total_AA_per_pos_JMJD6_substrate) ~
                        mean(
                            total_AA_per_pos_JMJD6_substrate,
                            na.rm = TRUE
                        ) %>% as.integer,
                    TRUE ~ total_AA_per_pos_JMJD6_substrate
                )]

odds.ratio.dt[
   , binom_p := binom.test(
        x = N_JMJD6_substrate, n = total_AA_per_pos_JMJD6_substrate,
        p = N_Others / total_AA_per_pos_Others
    )$p.value,
    by = seq_len(nrow(odds.ratio.dt))
]


odds.ratio.dt[, `:=`(
    adj_binom_p = p.adjust(binom_p, method = "fdr"),
    log2_odds_ratio = log2(odds_JMJD6_substrate / odds_Others)
)]

odds.ratio.dt[
  , capped_log2_odds_ratio := case_when(
        is.infinite(log2_odds_ratio) ~ sign(log2_odds_ratio) * 5,
        TRUE ~ log2_odds_ratio
)]

aa.order.dt <-
    odds.ratio.dt[
      , list(sum_odds = sum(capped_log2_odds_ratio, na.rm = TRUE)), by = amino_acid
    ][
        order(sum_odds)
    ]

odds.ratio.dt[, `:=`(
    amino_acid = factor(amino_acid, levels = aa.order.dt[, amino_acid]),
    position = gsub("position_", "", position2KOH) %>%
        {gsub("m", "-", .)} %>%
        factor(levels = -10:10)
)]

ggplot(
    data = odds.ratio.dt,
    aes(
        x = position,
        y = amino_acid,
        fill = capped_log2_odds_ratio#,
        ##color = adj_binom_p < 0.1
    )
) +
    geom_tile(size = 0.85) +
    scale_fill_gradient2(name = "Odds ratio") +
    scale_color_manual(values = c("TRUE" = "purple", "FALSE" = NA), name = "p(adj) < 0.1") +
    ylab("Amino acid") + 
    theme(
        legend.position = "bottom"
    )

```


## Biophysical properties


```{r analysis of sequence feature, fig.width = 6, fig.height = 6}

data.cols <- c("IUPRED2", "WindowHydropathy", "windowCharge", "K_ratio")

m.stoichiometry.dt <- melt(
    non.duplicated.stoichiometry.dt,
    measure.vars = data.cols,
    id.vars = c("cell", "Accession", "position", "JMJD6_substrate_flag", "ox_ratio")
)

m.stoichiometry.dt[, Z_score := scale(value), by = variable]


## ggplot(
##     data = non.duplicated.stoichiometry.dt,
##     aes(
##         x = windowCharge,
##         y = IUPRED2,
##         size = ox_ratio + 0.01,
##         fill = JMJD6_substrate_flag,
##         color = grepl("BRD", Accession)
##     )
## ) +
##     geom_jitter(shape = "circle filled", width = 0.04, height = 0) +
##     gghighlight::gghighlight(
##                      JMJD6_substrate_flag == "JMJD6_substrate",
##                      use_direct_label = FALSE,
##                      use_group_by = FALSE
##                  ) +
##     theme(
##         aspect.ratio = 1,
##         legend.position = "bottom",
##         legend.direction = "vertical"
##     ) +
##     scale_size_continuous(name = "Hydroxylation rate [%]") +
##     scale_fill_manual(values = c(
##                           "JMJD6_substrate" = "#EE6677",
##                           "Others" = "gray60"
##                       ), name = "JMJD6 substrate") +
##     scale_color_manual(values = c(
##                            "TRUE" = "black",
##                            "FALSE" = "gray60"
##                        ), name = "BRDs") +
##     xlab("Local charge") +
##     ylab("Disordedness") +
##     ggtitle("Biophysical propety around K")

ggplot(
    data = non.duplicated.stoichiometry.dt[order(JMJD6_substrate_flag != "Others")],
    aes(
        x = windowCharge,
        y = IUPRED2,
        size = ifelse(
            JMJD6_substrate_flag == "JMJD6_substrate",
            ox_ratio + 0.01,
            0.01
        ),
        fill = JMJD6_substrate_flag,
        alpha = JMJD6_substrate_flag,
        shape = JMJD6_substrate_flag
    )
) +
    geom_jitter(width = 0.04, height = 0) +
    theme(
        aspect.ratio = 1,
        legend.position = "bottom",
        legend.direction = "vertical"
    ) +
    scale_size_continuous(name = "Hydroxylation rate [%]") +
    scale_alpha_manual(
        values = c(
            "JMJD6_substrate" = 1,
            "Others" = 0.075
        ),
        name = "JMJD6 substrate"
    ) +
    scale_shape_manual(
        values = c(
            "JMJD6_substrate" = "circle filled",
            "Others" = "circle small"
        ),
        name = "JMJD6 substrate"
    ) +
    scale_fill_manual(values = c(
                          "JMJD6_substrate" = "#EE6677",
                          "Others" = "gray60"
                      ), name = "JMJD6 substrate") +
    scale_color_manual(values = c(
                           "TRUE" = "black",
                           "FALSE" = "gray60"
                       ), name = "BRDs") +
    xlab("Local charge") +
    ylab("Disordedness") +
    ggtitle("Biophysical propety around K")

## ggplot(
##     data = non.duplicated.stoichiometry.dt,
##     aes(
##         x = K_ratio_score,
##         y = IUPRED2,
##         size = ox_ratio + 0.01,
##         fill = JMJD6_substrate_flag,
##         color = grepl("BRD", Accession)
##     )
## ) +
##     geom_jitter(shape = "circle filled", width = 0.04, height = 0) +
##     gghighlight::gghighlight(
##                      JMJD6_substrate_flag == "JMJD6_substrate",
##                      use_direct_label = FALSE,
##                      use_group_by = FALSE
##                  ) +
##     theme(
##         aspect.ratio = 1,
##         legend.position = "bottom",
##         legend.direction = "vertical"
##     ) +
##     scale_size_continuous(name = "Hydroxylation rate [%]") +
##     scale_fill_manual(values = c(
##                           "JMJD6_substrate" = "#EE6677",
##                           "Others" = "gray60"
##                       ), name = "JMJD6 substrate") +
##     scale_color_manual(values = c(
##                            "TRUE" = "black",
##                            "FALSE" = "gray60"
##                        ), name = "BRDs") +
##     xlab("Local K ratio") +
##     ylab("Disordedness") +
##     ggtitle("Biophysical propety around K 2")

## ggplot(
##     data = non.duplicated.stoichiometry.dt,
##     aes(
##         x = K_ratio_score,
##         y = windowCharge,
##         size = ox_ratio + 0.01,
##         fill = JMJD6_substrate_flag,
##         color = grepl("BRD", Accession)
##     )
## ) +
##     geom_jitter(shape = "circle filled", width = 0.04, height = 0.04) +
##     gghighlight::gghighlight(
##                      JMJD6_substrate_flag == "JMJD6_substrate",
##                      use_direct_label = FALSE,
##                      use_group_by = FALSE
##                  ) +
##     theme(
##         aspect.ratio = 1,
##         legend.position = "bottom",
##         legend.direction = "vertical"
##     ) +
##     scale_size_continuous(name = "Hydroxylation rate [%]") +
##     scale_fill_manual(values = c(
##                           "JMJD6_substrate" = "#EE6677",
##                           "Others" = "gray60"
##                       ), name = "JMJD6 substrate") +
##     scale_color_manual(values = c(
##                            "TRUE" = "black",
##                            "FALSE" = "gray60"
##                        ), name = "BRDs") +
##     xlab("Local K ratio") +
##     ylab("Local charge") +
##     ggtitle("Biophysical propety around K 3")


## melt(
##     non.duplicated.stoichiometry.dt,
##     id.vars = c("Accession", "position", "JMJD6_substrate_flag"),
##     measure.vars = c("windowCharge", "K_ratio_score")
## ) %>%
##     {.[, z_score := scale(value), by = variable]} %>%
##     ggplot(
##         aes(
##             x = variable,
##             y = z_score,
##             fill = JMJD6_substrate_flag
##         )
##     ) +
##     geom_boxplot(outlier.shape = NA) +
##     scale_fill_bright()


```


```{r plot all}

## library("plotly")

## non.duplicated.stoichiometry.dt[, `:=`(
##     j_windowCharge = jitter(windowCharge, factor = 1, amount = 0.025),
##     j_K_ratio_score = jitter(K_ratio_score, factor = 1, amount = 0.025),
##     plot_flag = case_when(
##         JMJD6_substrate_flag == "JMJD6_substrate" ~ TRUE,
##         TRUE ~ sample(
##             rep(c(TRUE, FALSE), times = c(1, 999)),
##             size = nrow(non.duplicated.stoichiometry.dt),
##             replace = TRUE
##         )
##     )
## )]

## plot_ly(
##     data = non.duplicated.stoichiometry.dt[plot_flag == TRUE],
##     x = ~ j_windowCharge,
##     y = ~ j_K_ratio_score,
##     z = ~ IUPRED2,
##     ## size = ~ ox_ratio + 0.01,
##     color = ~ JMJD6_substrate_flag,
##     type = "scatter3d"
## )

```


# Session information


```{r session info}

sessioninfo::session_info()


```
