---
title: "Analysis of lysine hydroxylation stoichiometry and protein features"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteIndexEntry{Bioconductor style for PDF documents}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   github_document:
     toc: yes
     toc_depth: 2
---


# Package import


```{r load packages, message = FALSE, warning = FALSE}

temp <- sapply(list.files("../functions", full.names = TRUE), source)

library("Biostrings")
library("GenomicRanges")
library("BSgenome")
library("readxl")

processors <- 8

set.seed(1)

```


```{r dir create}

j3.input.dir <- file.path("../../data/j3-properties-of-hydroxylation-sites")

results.dir <- normalizePath(file.path("../../results"))
j0.res.dir <- file.path(results.dir, "j0-data-preprocessing")
j2.res.dir <- file.path(results.dir, "j2-PTM-stoichiometry")
j2.1.res.dir <- file.path(j2.res.dir, "j2-1-PTM-stoichiometry-K-only")

create.dirs(c(
))

```


# Data import

## Annotation


```{r annotation}

all.protein.bs <- readAAStringSet("../../data/all_protein.fasta")
names(all.protein.bs) <- names(all.protein.bs) %>%
    {gsub("sp\\|", "", .)} %>%
    {str_split_fixed(., " ", n = 2)[, 1]}
uniprot.accession.flag <- names(all.protein.bs) %>%
    {str_split_fixed(., "\\|", n = 2)[, 2]} %>%
    {nchar(.) > 0}
all.protein.bs <- all.protein.bs[uniprot.accession.flag]

all.protein.feature.per.pos.dt <- file.path(
    j0.res.dir, "all_protein_feature_per_position.csv"
) %>% fread

```


## Stoichiometry data

```{r stoichiometry_data, fig.width = 10, fig.height = 5}

stoichiometry.dt <- fread(
    file.path(
        j2.1.res.dir,
        "long_K_stoichiometry_data.csv"
    )
)

stoichiometry.dt[
  , curated_oxK_site := case_when(
        curated_oxK_site == TRUE ~ "JMJD6_substrate",
        TRUE ~ "Others"
    )
]

stoichiometry.dt[, `:=`(
    stoichiometry_available = case_when(
        is.na(oxK_ratio) ~ FALSE,
        TRUE ~ TRUE
    ),
    Accession_position = paste(Accession, position, sep = "_")
)]

## Sanity check
j6pep.substrate.dt <- stoichiometry.dt[
    stoichiometry_available == TRUE &
    curated_oxK_site == "JMJD6_substrate" &
    data_source == "HeLa_WT_J6pep" &
    total_n_feature_K > 1 &
    grepl("J6pep", screen),
    .(Accession, screen, position, total_n_feature_K, oxK_ratio)
][order(oxK_ratio, decreasing = TRUE)]

j6pep.substrate.dt[
  , accession_position := str_split_fixed(Accession, "\\|", n = 2)[, 2] %>%
        {gsub("_HUMAN", "", .)} %>%
        {paste0(., " (", position, ")")} %>%
        {factor(., levels = .)}
]

ggplot(
    data = j6pep.substrate.dt,
    aes(
        x = accession_position,
        y = oxK_ratio * 100
    )
) +
    geom_bar(stat = "identity") +
    coord_cartesian(ylim = c(0, 100)) +
    theme(
        axis.title.x = element_blank()
    ) +
    scale_x_discrete(guide = guide_axis(angle = 90)) +
    ylab("Stoichiometry [%]") +
    ggtitle("JMJD6 peptide pulldown data")

reportOxKSiteStats <- function(stoichiometry.dt){
    print("Total unique JMJD6 substrate proteins and sites:")
    stoichiometry.dt[, list(
        unique_protein_N = uniqueN(Accession),
        unique_site_N = uniqueN(Accession_position)
    ),
    by = list(
        curated_oxK_site
    )
    ][order(curated_oxK_site)] %>%
        print

    print("by the availability of stocihiometry data")
    stoichiometry.dt[, list(
        unique_protein_N = uniqueN(Accession),
        unique_site_N = uniqueN(Accession_position)
    ),
    by = list(
        curated_oxK_site,
        stoichiometry_available
    )
    ][order(curated_oxK_site, -stoichiometry_available)] %>%
        print

    print("by the data source")
    stoichiometry.dt[, list(
        unique_protein_N = uniqueN(Accession),
        unique_site_N = uniqueN(Accession_position)
    ),
    by = list(
        curated_oxK_site,
        stoichiometry_available,
        data_source
    )
    ][order(curated_oxK_site, data_source)] %>%
        print

    return()
}

temp <- reportOxKSiteStats(stoichiometry.dt)

print("Hydroxylation sites identified by non-unique peptides; see the main text.")
stoichiometry.dt[
    order(Accession, position, -oxK_ratio)
][, head(.SD, 1), by = list(Accession, position)][
    curated_oxK_site == "JMJD6_substrate" &
    (stoichiometry_available == FALSE | oxK_ratio == 0),
    .(Accession, position, curated_oxK_site, screen)
]

```

# Analysis of the properties of hydroxylation sites


```{r fasta_data}

stoichiometry.dt[, `:=`(
    protein_sequence = BSgenome::getSeq(
                             x = all.protein.bs,
                             name = Accession
                         ) %>% as.character
)]

seq.search.range <- 10

stoichiometry.dt[, `:=`(
    sequence = substr(
        protein_sequence,
        start = pmax(0, position - seq.search.range),
        stop = position + seq.search.range
    ),
    protein_sequence = NULL
)]

stoichiometry.dt[, full_length_seq_flag := (
    nchar(sequence) == (seq.search.range * 2 + 1)
)]

left.cols <- c(
    "data_source", "Accession", "uniprot_id", "position", "sequence",
    "full_length_seq_flag",
    "curated_oxK_site",
    "oxK_ratio"
)

stoichiometry.dt <- stoichiometry.dt[, c(
    left.cols, colnames(stoichiometry.dt)[!(colnames(stoichiometry.dt) %in% left.cols)]
), with = FALSE]

## Additional data filteration by the number of feature, and the availability of full length sequence

stoichiometry.dt <- stoichiometry.dt[
    curated_oxK_site == "JMJD6_substrate" |
    (
        residue == "K" &
        full_length_seq_flag == TRUE &
        total_n_feature_K > 2
    )
]

temp <- reportOxKSiteStats(stoichiometry.dt)

```

# Analysis of sequence feature around hydroxylation site

## Amino acid enrichment around the hydroxylation sites


```{r amino_acid_enrichment, fig.width = 7}

print("the following sites were excluded from the analysis:")
stoichiometry.dt[
    (data_source %in% c("HeLa_WT_JQ1", "HeLa_WT_J6pep")) |
    (curated_oxK_site == "JMJD6_substrate")
][CenterResidue != "K"][
    order(
        data_source %in% c("HeLa_WT_JQ1", "HeLa_WT_J6pep"),
        curated_oxK_site == "JMJD6_substrate",
        oxK_ratio,
        decreasing = TRUE
    )
][
    !duplicated(paste(Accession_position))
]

non.duplicated.stoichiometry.dt <- stoichiometry.dt[
    (data_source %in% c("HeLa_WT_JQ1", "HeLa_WT_J6pep")) |
    (curated_oxK_site == "JMJD6_substrate")
][CenterResidue == "K"][
    order(
        data_source %in% c("HeLa_WT_JQ1", "HeLa_WT_J6pep"),
        curated_oxK_site == "JMJD6_substrate",
        oxK_ratio,
        decreasing = TRUE
    )
][
    !duplicated(paste(Accession_position))
]

temp <- reportOxKSiteStats(non.duplicated.stoichiometry.dt)

## Sanity checks
non.duplicated.stoichiometry.dt[duplicated(Accession_position)]
non.duplicated.stoichiometry.dt[, table(CenterResidue)]

## Back to analysis
non.duplicated.stoichiometry.dt[
  , (paste0("position_", gsub("-", "m", seq(-10, 10)))) :=
        str_split(sequence, pattern = "")[[1]] %>% as.list,
    by = seq_len(nrow(non.duplicated.stoichiometry.dt))
]

key.id.cols <- c("data_source", "Accession", "position", "curated_oxK_site", "oxK_ratio")

m.non.duplicated.stoichiometry.dt <- melt(
    non.duplicated.stoichiometry.dt,
    id.vars = c(key.id.cols),
    measure.vars = paste0("position_", gsub("-", "m", seq(-10, 10))),
    variable.name = "position2KOH",
    value.name = "amino_acid"
)

aa.count.per.pos.dt <- m.non.duplicated.stoichiometry.dt[
  , .N, by = list(curated_oxK_site, position2KOH, amino_acid)
]

aa.count.per.pos.dt[
  , total_AA_per_pos := sum(N),
    by = list(curated_oxK_site, position2KOH)
]

aa.count.per.pos.dt[
  , odds := (N / total_AA_per_pos) %>% {./(1 - .)}
]

odds.ratio.dt <- dcast(
    aa.count.per.pos.dt,
    position2KOH + amino_acid ~ curated_oxK_site,
    value.var = c("odds", "N", "total_AA_per_pos")
)

setnafill(
    odds.ratio.dt, fill = 0, cols = c("odds_JMJD6_substrate", "N_JMJD6_substrate")
)

odds.ratio.dt[, total_AA_per_pos_JMJD6_substrate := case_when(
                    is.na(total_AA_per_pos_JMJD6_substrate) ~
                        mean(
                            total_AA_per_pos_JMJD6_substrate,
                            na.rm = TRUE
                        ) %>% as.integer,
                    TRUE ~ total_AA_per_pos_JMJD6_substrate
                )]

odds.ratio.dt[
   , binom_p := binom.test(
        x = N_JMJD6_substrate, n = total_AA_per_pos_JMJD6_substrate,
        p = N_Others / total_AA_per_pos_Others
    )$p.value,
    by = seq_len(nrow(odds.ratio.dt))
]


odds.ratio.dt[, `:=`(
    adj_binom_p = p.adjust(binom_p, method = "fdr"),
    log2_odds_ratio = log2(odds_JMJD6_substrate / odds_Others)
)]

odds.ratio.dt[
  , capped_log2_odds_ratio := case_when(
        is.infinite(log2_odds_ratio) ~ sign(log2_odds_ratio) * 5,
        TRUE ~ log2_odds_ratio
)]

aa.order.dt <-
    odds.ratio.dt[
      , list(sum_odds = sum(capped_log2_odds_ratio, na.rm = TRUE)), by = amino_acid
    ][
        order(sum_odds)
    ]

odds.ratio.dt[, `:=`(
    amino_acid = factor(amino_acid, levels = aa.order.dt[, amino_acid]),
    position = gsub("position_", "", position2KOH) %>%
        {gsub("m", "-", .)} %>%
        factor(levels = -10:10)
)]

ggplot(
    data = odds.ratio.dt,
    aes(
        x = position,
        y = amino_acid,
        fill = capped_log2_odds_ratio#,
        ##color = adj_binom_p < 0.1
    )
) +
    geom_tile(size = 0.85) +
    scale_fill_gradient2(name = "Odds ratio") +
    scale_color_manual(values = c("TRUE" = "purple", "FALSE" = NA), name = "p(adj) < 0.1") +
    ylab("Amino acid") + 
    theme(
        legend.position = "bottom"
    )

```


## Biophysical properties


```{r analysis_of_sequence_feature, fig.width = 6, fig.height = 6}

non.duplicated.stoichiometry.dt <-
    non.duplicated.stoichiometry.dt[
        data_source %in% c("HeLa_WT_JQ1", "HeLa_WT_J6pep") &
        total_n_feature_K > 2 &
        !is.na(windowCharge)
    ]

temp <- reportOxKSiteStats(non.duplicated.stoichiometry.dt)

## Sanity checks
non.duplicated.stoichiometry.dt[, table(CenterResidue)]

data.cols <- c("IUPRED2", "WindowHydropathy", "windowCharge", "K_ratio")

## non.duplicated.stoichiometry.dt[
##     curated_oxK_site == "JMJD6_substrate"
## ][order(windowCharge)][, .(
##      Accession, position, Window, windowCharge, IUPRED2
##  )]


ggplot(
    data = non.duplicated.stoichiometry.dt[order(curated_oxK_site != "Others")],
    aes(
        x = windowCharge,
        y = IUPRED2,
        size = ifelse(
            curated_oxK_site == "JMJD6_substrate",
            oxK_ratio + 0.01,
            0.01
        ),
        fill = curated_oxK_site,
        alpha = curated_oxK_site,
        shape = curated_oxK_site
    )
) +
    geom_jitter(width = 0.04, height = 0) +
    theme(
        aspect.ratio = 1,
        legend.position = "bottom",
        legend.direction = "vertical"
    ) +
    scale_size_continuous(
        name = "Hydroxylation rate [%]"
    ) +
    scale_alpha_manual(
        values = c(
            "JMJD6_substrate" = 1,
            "Others" = 0.075
        ),
        name = "JMJD6 substrate"
    ) +
    scale_shape_manual(
        values = c(
            "JMJD6_substrate" = "circle filled",
            "Others" = "circle small"
        ),
        name = "JMJD6 substrate"
    ) +
    scale_fill_manual(values = c(
                          "JMJD6_substrate" = "#EE6677",
                          "Others" = "gray60"
                      ), name = "JMJD6 substrate") +
    scale_color_manual(values = c(
                           "TRUE" = "black",
                           "FALSE" = "gray60"
                       ), name = "BRDs") +
    xlab("Local charge") +
    ylab("Disorderedness") +
    ggtitle("Biophysical propety around K")


print("Test for disorderedness")
non.duplicated.stoichiometry.dt %$%
    {wilcox.test(
         .[curated_oxK_site == "JMJD6_substrate", IUPRED2],
         .[curated_oxK_site == "Others", IUPRED2]
    )}

print("Test for local charge")
non.duplicated.stoichiometry.dt %$%
    {wilcox.test(
         .[curated_oxK_site == "JMJD6_substrate", windowCharge],
         .[curated_oxK_site == "Others", windowCharge]
    )}


```



# Session information


```{r session_info}

sessioninfo::session_info()


```
